### PR TITLE
Updated to have error codes in the services

### PR DESCRIPTION
**Pull request by bgill92**
## Description

The service messages now have error codes.

## Testing

In the docker container run 

```
colcon build
colcon test
```
Make sure there are no test failures

### CODE REVIEW

**Comment by griswaldbrooks** [comment]
looks like you forgot to add the new messages

File: functional_programming_tests/include/pathing/pathing.hpp [hunk 1]
```diff
@@ -74,10 +74,10 @@ namespace generate_path {
  * @brief      Types of errors expected in the generate path callback function
  */
 enum class error {
-  EMPTY_OCCUPANCY_MAP,
-  INVALID_START_SIZE,
-  INVALID_GOAL_SIZE,
-  NO_VALID_PATH
+  EMPTY_OCCUPANCY_MAP = 1,
```
Comment (griswaldbrooks) [inline functional_programming_tests/include/pathing/pathing.hpp]: is there anyway you can include the message files and do, something like
```
EMPTY_OCCUPANCY_MAP = GetPathCodes::msg::EMPTY_OCCUPANCY_MAP
```
idk what the actual constant is named or if this is even possible.
I've just noticed in past projects that having these two sources of truth can cause issues

File: functional_programming_tests/src/pathing.cpp [hunk 2]
```diff
@@ -115,7 +115,7 @@ tl::expected<example_srvs::srv::GetPath::Response, error> generate_path(
   }
 
   auto response = example_srvs::srv::GetPath::Response{};
-  response.success.data = path.has_value();
+  response.code.code = 0;
```
Comment (griswaldbrooks) [inline functional_programming_tests/src/pathing.cpp]: I would add a success message to the enum/message file and use that so we don't use numbers and stick with semantics

File: functional_programming_tests/src/pathing_manager.cpp [hunk 3]
```diff
@@ -18,33 +19,62 @@ Manager::Manager(std::unique_ptr<Manager::MiddlewareHandle> mw)
...
-      response.success.data = false;
+      response.code.code = static_cast<int>(error);
       response.path = std_msgs::msg::UInt8MultiArray();
       return response;
     };
-    auto const stringify_error = [](auto const error) {
-      return generate_path::error_description.at(error);
-    };
+
+    std::cerr << "In path callback, map size is : " << this->map_.get_data().size() << "\n";
```
Comment (griswaldbrooks) [inline functional_programming_tests/src/pathing_manager.cpp]: still need this? or should you use `mw_->log_error` here so the side effects are captured

File: functional_programming_tests/src/pathing_manager.cpp [hunk 4]
```diff
@@ -18,33 +19,62 @@ Manager::Manager(std::unique_ptr<Manager::MiddlewareHandle> mw)
...
       [this](auto const request, auto response) -> void {
-        auto const path = utilities::parseSetMapRequest(request);
-        if (path) this->map_ = path.value();
-        response->success.data = path.has_value();
+
+        auto const set_map = [this](Map<unsigned char> const map)
+        -> tl::expected<Map<unsigned char>, utilities::parsing_set_map_error> {
+          this->map_ = map;
+          return map;
+        };
```
Comment (griswaldbrooks) [inline functional_programming_tests/src/pathing_manager.cpp]: I think you can just return `void`
```suggestion
        auto const set_map = [this](Map<unsigned char> const map)
        -> tl::expected<void, utilities::parsing_set_map_error> {
          this->map_ = map;
          return;
        };
```

File: functional_programming_tests/src/pathing_manager.cpp [hunk 5]
```diff
@@ -18,33 +19,62 @@ Manager::Manager(std::unique_ptr<Manager::MiddlewareHandle> mw)
...
+
+        auto const return_successful_response = []([[maybe_unused]] auto const success)
```
Comment (griswaldbrooks) [inline functional_programming_tests/src/pathing_manager.cpp]:
```suggestion
        auto const return_successful_response = []([[maybe_unused]] auto const) 
```

File: functional_programming_tests/src/pathing_manager.cpp [hunk 6]
```diff
@@ -18,33 +19,62 @@ Manager::Manager(std::unique_ptr<Manager::MiddlewareHandle> mw)
...
+        auto const set_map = [this](Map<unsigned char> const map)
+        -> tl::expected<Map<unsigned char>, utilities::parsing_set_map_error> {
+          this->map_ = map;
+          return map;
+        };
+
+        auto const return_successful_response = []([[maybe_unused]] auto const success) 
+        -> tl::expected<SetMap::Response, utilities::parsing_set_map_error> {
+          auto response = SetMap::Response{};
+          response.code.code = 0;
```
Comment (griswaldbrooks) [inline functional_programming_tests/src/pathing_manager.cpp]: same comment about using the named variable from the error code message or enum

File: functional_programming_tests/src/pathing_manager.cpp [hunk 7]
```diff
@@ -18,33 +19,62 @@ Manager::Manager(std::unique_ptr<Manager::MiddlewareHandle> mw)
...
+          return response;
+        };
+
+        auto const print_error = [this](utilities::parsing_set_map_error const error)
+            -> tl::expected<SetMap::Response, utilities::parsing_set_map_error> {
+          mw_->log_error(std::string{utilities::parsing_set_map_error_description.at(error)});
```
Comment (griswaldbrooks) [inline functional_programming_tests/src/pathing_manager.cpp]: does
```suggestion
          mw_->log_error(utilities::parsing_set_map_error_description.at(error));
```
work?

File: functional_programming_tests/src/pathing_manager.cpp [hunk 8]
```diff
@@ -18,33 +19,62 @@ Manager::Manager(std::unique_ptr<Manager::MiddlewareHandle> mw)
...
       });
   mw_->register_generate_path_service([this](auto const request,
                                              auto response) {
-    auto const print_error = [this](std::string_view error)
-        -> tl::expected<GetPath::Response, std::string> {
-      mw_->log_error(std::string{error});
-      return tl::make_unexpected("");
+    auto const print_error = [this](auto const error)
+        -> tl::expected<GetPath::Response, pathing::generate_path::error> {
+      mw_->log_error(std::string{pathing::generate_path::error_description.at(error)});
```
Comment (griswaldbrooks) [inline functional_programming_tests/src/pathing_manager.cpp]:
```suggestion
      mw_->log_error(pathing::generate_path::error_description.at(error));
```
?

File: example_srvs/srv/GetPath.srv [hunk 9]
```diff
@@ -1,5 +1,5 @@
 std_msgs/UInt8MultiArray start
 std_msgs/UInt8MultiArray goal
 ---
-std_msgs/Bool success
+example_srvs/GetPathCodes code
```
Comment (griswaldbrooks) [inline example_srvs/srv/GetPath.srv]: maybe this should be
```suggestion
example_srvs/GetPathCodes result
```
so latter the setter is
```
result.code = ...
```

**Review by griswaldbrooks — CHANGES_REQUESTED** [review]
First pass

File: functional_programming_tests/src/pathing_manager.cpp (hunk 3 above)
Comment (bgill92) [reply functional_programming_tests/src/pathing_manager.cpp]: I think that was an artifact of me debugging, removing.

File: functional_programming_tests/src/pathing_manager.cpp (hunk 4 above)
Comment (bgill92) [reply functional_programming_tests/src/pathing_manager.cpp]: Nope, compiling fails

File: functional_programming_tests/src/pathing_manager.cpp (hunk 7 above)
Comment (bgill92) [reply functional_programming_tests/src/pathing_manager.cpp]: Yee

**Review by griswaldbrooks — DISMISSED** [review]

**Review by griswaldbrooks — APPROVED** [review]

**Pull request merged**